### PR TITLE
Fix layout validators to reject negative percentages in spacing props

### PR DESCRIPTION
**Pull request by Faye-Xiao**
## Summary

- `isPercentage` in `tokens.ts` used `-?` in its regex, unintentionally allowing negative percentages to pass validation for `padding`, `margin`, and `gap` props
- Negative percentages are not meaningful CSS spacing values and should be rejected
- `isValidPositionValue` and `isValidSizeValue` intentionally support negative values (e.g. `top: '-50%'` for translate offsets), so they now inline their own `/^-?\d+(\.\d+)?%$/` regex instead of delegating to `isPercentage`

**Files changed:**
- `tokens.ts` — remove `-?` from `isPercentage`; inline negative-% regex in `isValidPositionValue` and `isValidSizeValue`
- `tokenUtils-test.ts` — add tests asserting negative % is rejected for `padding`, `margin`, `gap` and accepted for position props

## Test plan

- [ ] `pnpm run jest --testPathPattern="tokenUtils-test"` — 41 tests pass
- [ ] `eslint` on changed files — no errors
- [ ] Negative % for `padding`/`margin`/`gap` is dropped with a console warning
- [ ] Negative % for `top`/`left`/`right`/`bottom` continues to pass through unchanged

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts [hunk 1]
```diff
@@ -200,25 +202,27 @@ export function isValidMarginValue(value: string): boolean {
  */
 export function isValidSizeValue(value: string): boolean {
   return (
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokens.ts:208]: `isValidSizeValue` comment is misleading (likely a pre-existing bug surfaced here)

```
/^-?\d+(\.\d+)?rem$/.test(value) || // rem values (negative allowed for transforms)
/^-?\d+(\.\d+)?%$/.test(value) ||   // percentage values (negative allowed)
```
`isValidSizeValue` governs width, height, minWidth, maxWidth, etc. Negative values for these properties are invalid CSS — a negative width or minWidth is meaningless. The comment "negative allowed for transforms" doesn't make sense in this context (transform uses isValidTransformValue, not this function). This looks like a copy-paste from the position validator. Worth either:

- Dropping the -? from both lines here (if negative size was never intentional), or
- Adding a note confirming this is intentional with a concrete use-case.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts [hunk 2]
```diff
@@ -200,25 +202,27 @@ export function isValidMarginValue(value: string): boolean {
...
-    /^-?\d+(\.\d+)?rem$/.test(value) || // rem values
-    isPercentage(value) || // percentage values
+    /^-?\d+(\.\d+)?rem$/.test(value) || // rem values (negative allowed for transforms)
+    /^-?\d+(\.\d+)?%$/.test(value) || // percentage values (negative allowed)
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokens.ts]: The `regex /^-?\d+(\.\d+)?%$/` now appears in two places. A named export like:


`const NEGATIVE_PERCENTAGE_RE = /^-?\d+(\.\d+)?%$/;`
would remove the duplication and make the intent self-documenting. Low priority, but worth doing while the change is open.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts [hunk 3]
```diff
@@ -431,6 +431,46 @@ describe('processBpkProps', () => {
...
+  });
+
+  it('rejects negative percentage for gap and drops the prop', () => {
+    const warnSpy = jest.spyOn(console, 'warn').mockImplementation(() => {});
+
+    const result = processBpkProps({
+      gap: '-10%' as any,
+    });
+
+    expect(result.gap).toBeUndefined();
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts:484]: BpkFlex exposes rowGap and columnGap as separate gap props. The new "rejects negative percentage for gap" test only covers gap. If those props go through a different validator path, the fix might not cover them. Worth adding:


```
const result = processBpkProps({ gap: '-10%', rowGap: '-5%', columnGap: '-5%' } as any);
expect(result.rowGap).toBeUndefined();
expect(result.columnGap).toBeUndefined();
```

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts [hunk 4]
```diff
@@ -431,6 +431,46 @@ describe('processBpkProps', () => {
...
+    expect(warnSpy).toHaveBeenCalled();
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts]: All three new tests assert only expect(warnSpy).toHaveBeenCalled(). If the warning message changes to something uninformative (or is removed), this passes silently. Prefer:


expect(warnSpy).toHaveBeenCalledWith(expect.stringContaining('-10%'));
or at minimum toHaveBeenCalledTimes(1) per rejected prop so the count is explicit.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts [hunk 5]
```diff
@@ -422,7 +422,7 @@ describe('processBpkProps', () => {
     expect(result.scrollMarginBottom).toBe('0');
   });
 
-  it('passes through negative percentages for position props (e.g. top: -50% for vertical centering)', () => {
+  it('passes through negative percentages for position props (e.g. top: -50% for translate offsets)', () => {
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts:429]: There's no test confirming that padding: '50%' still passes after the change. Given this is the exact behavior the fix changes, one positive assertion is cheap insurance.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts [hunk 6]
```diff
@@ -200,25 +202,27 @@ export function isValidMarginValue(value: string): boolean {
...
     value === 'auto' ||
     value === 'full' ||
     value === 'fit-content'
```
[the remark this answers is not in the document]
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokens.ts:213]: Fixed in the follow-up commit — `isValidSizeValue` now uses `/^\d+(\.\d+)?rem$/` (no `-?`) and delegates to `isPercentage` (non-negative) instead of inlining the signed regex.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts [hunk 7]
```diff
@@ -200,25 +202,27 @@ export function isValidMarginValue(value: string): boolean {
  */
 export function isValidSizeValue(value: string): boolean {
   return (
-    /^-?\d+(\.\d+)?rem$/.test(value) || // rem values
-    isPercentage(value) || // percentage values
+    /^-?\d+(\.\d+)?rem$/.test(value) || // rem values (negative allowed for transforms)
```
[the remark this answers is not in the document]
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokens.ts]: Fixed — the misleading comment and the signed regex in `isValidSizeValue` were both removed in the follow-up commit.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts [hunk 8]
```diff
@@ -431,6 +431,46 @@ describe('processBpkProps', () => {
...
+    const result = processBpkProps({
+      padding: '-10%' as any,
+      paddingTop: '-5%' as any,
+    });
+
+    expect(result.padding).toBeUndefined();
+    expect(result.paddingTop).toBeUndefined();
+    expect(warnSpy).toHaveBeenCalled();
+    warnSpy.mockRestore();
+  });
```
[the remark this answers is not in the document]
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts:458]: Fixed — added `afterEach(() => jest.restoreAllMocks())` to the `processBpkProps` describe block so spy cleanup is guaranteed even if an assertion throws.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts (hunk 1 above)
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokens.ts:208]: Fixed — `isValidSizeValue` now uses `/^\d+(\.\d+)?rem$/` (no `-?`) and `isPercentage` (non-negative). Negative size values were never intentional; this was a copy-paste from the position validator.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts (hunk 2 above)
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokens.ts]: Done — extracted as `NEGATIVE_PERCENTAGE_RE` and used in `isValidPositionValue` to eliminate the duplication.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts (hunk 3 above)
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts:484]: Done — expanded the gap test to cover `rowGap` and `columnGap` as well.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts (hunk 4 above)
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts]: Done — assertions now use `toHaveBeenCalledWith(expect.stringContaining(value))` so they verify the invalid value appears in the warning message.

File: packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts (hunk 5 above)
Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokenUtils-test.ts:429]: Done — added a dedicated test: `padding: '50%'` passes through as `'50%'` after the change.

File: packages/backpack-web/src/bpk-component-layout/src/tokens.ts [hunk 9]
```diff
@@ -161,21 +161,27 @@ export type BpkResponsiveValue<T> =
   | T
   | Partial<Record<BpkBreakpointToken | 'base', T>>;
 
+// Matches percentage strings that include a leading minus sign, e.g. '-50%', '-0.5%'.
+// Used only by position validators where negative offsets are intentional.
+const NEGATIVE_PERCENTAGE_RE = /^-?\d+(\.\d+)?%$/;
```
Comment (Supremeyh) [inline packages/backpack-web/src/bpk-component-layout/src/tokens.ts]: how about rename to `SIGNED_PERCENTAGE_RE` as `-?` let this regex match both positive and negative percentages simultaneously (both 50% and -50% can pass)

Also comment should update
```
// Matches both positive and negative percentage strings, e.g. '50%', '-50%', '-0.5%'.
// Used by position validators where negative offsets are intentional.
const SIGNED_PERCENTAGE_RE = /^-?\d+(\.\d+)?%$/;

```

Comment (Faye-Xiao) [reply packages/backpack-web/src/bpk-component-layout/src/tokens.ts]: Good point — renamed to `SIGNED_PERCENTAGE_RE` and updated the comment accordingly.

**Review by Supremeyh — APPROVED** [review]
